### PR TITLE
General react performance improvements

### PR DESCRIPTION
**Pull request by jwilander**

### CODE REVIEW

File: webapp/components/needs_team.jsx [hunk 1]
```diff
@@ -42,20 +42,17 @@ export default class NeedsTeam extends React.Component {
         this.onChanged = this.onChanged.bind(this);
 
         this.state = {
-            profiles: UserStore.getProfiles(),
```
Comment (coreyhulen) [inline webapp/components/needs_team.jsx]: What happens if Profiles change after TeamsChange?  then the component will not re-render?

I switch teams.  fires getMyTeam and getProfiles(forNewTeamId) and getMyTeam finishes before getProfiles, won't it be showing the old profiles?

Comment (jwilander) [inline webapp/components/needs_team.jsx]: Yes, this component will not render again in that case but it doesn't need to since the only child that used the profiles was PostsView, which will still re-render when it gets an event about profiles changing. I just moved this code down into PostsView since there was no reason to have it this high up. None of the other children of NeedsTeam were using the profiles.

Having the profiles here was causing unneeded rendering of NeedsTeam which was really slow since it's such a high-level component.

**Comment by coreyhulen** [comment]
+1

File: webapp/components/more_channels.jsx [hunk 2]
```diff
@@ -33,6 +36,8 @@ export default class MoreChannels extends React.Component {
         this.handleNewChannel = this.handleNewChannel.bind(this);
         this.createChannelRow = this.createChannelRow.bind(this);
 
+        this.shouldComponentUpdate = PureRenderMixin.shouldComponentUpdate.bind(this);
```
Comment (crspeller) [inline webapp/components/more_channels.jsx]: Technically this one isn't pure because it depends on render height. However it is already broken so I think it is fine.

File: webapp/components/posts_view_container.jsx [hunk 3]
```diff
@@ -60,17 +59,14 @@ export default class PostsViewContainer extends React.Component {
         ChannelStore.addLeaveListener(this.onChannelLeave);
         PostStore.addChangeListener(this.onPostsChange);
         PostStore.addPostsViewJumpListener(this.handlePostsViewJumpRequest);
-        UserStore.addChangeListener(this.onUserChange);
+        $('body').addClass('app__body');
```
Comment (crspeller) [inline webapp/components/posts_view_container.jsx]: This seems unrelated.

Comment (jwilander) [inline webapp/components/posts_view_container.jsx]: It was being done inside post_views.jsx which was causing it to be applied and unapplied on each mount/unmount of one of those. I just put it here so it's added once and removed once.

**Comment by crspeller** [comment]
+1

File: webapp/components/posts_view_container.jsx (hunk 3 above)
Comment (hmhealey) [inline webapp/components/posts_view_container.jsx]: This should actually be removed from the posts_view_container. I moved it to the channel_view in https://github.com/mattermost/platform/pull/2763
